File: plugins/push/android/src/main/java/io/letary/lecodes/plugins/push/PushService.kt
//
//  PushService.kt — lecodes-plugins/push
//
//  Per-session adapter for the "push" service (registerService — docs/push-plan.md). Method
//  dispatch only; all state lives in PushManager. getStatus deliberately works without a
//  project identity and never prompts — the SDK's Push.addEventListener pokes the session
//  open with it to establish the event pipe ("message"/"tap" ride Services.Channel.emit).
//

package io.letary.lecodes.plugins.push

import io.letary.lecodes.utils.Services

class PushService(private val channel: Services.Channel) {

    init {
        PushManager.addSession(channel)
    }

    fun instance() = Services.Instance(
        onCall = { call ->
            when (call.method) {
                "getStatus" -> call.resolve(PushManager.getStatus())
                "register" -> PushManager.register(
                    call,
                    call.args.optJSONObject(0)?.optString("user")?.takeIf { it.isNotEmpty() },
                )
                "unregister" -> PushManager.unregister(call)
                "getLaunch" -> call.resolve(PushManager.getLaunch())
                else -> call.reject("push: unknown method \"${call.method}\"")
            }
        },
        onClose = { PushManager.removeSession(channel) },
    )
}

File: plugins/push/ios/PushService.swift
//
//  PushService.swift — lecodes-plugins/push
//
//  The per-session "push" service adapter (docs/push-plan.md; the Geolocation recipe
//  applied to notifications):
//    call getStatus()        → { permission: "granted"|"denied"|"prompt", registered } — never prompts
//    call register({user?})  → { address }; OS permission prompt on first use;
//                              rejects "denied" / "unavailable"
//    call unregister()       → removes the registration (backend + local)
//    call getLaunch()        → the notification payload that cold-started this world, or null
//    event "message"         → payload arrived while this app was foregrounded (banner suppressed)
//    event "tap"             → payload tapped while this world was alive (warm)
//
//  Sessions open cheaply (no prompt, no I/O — the channel contract); every call resolves
//  the CURRENT world's identity at call time.
//

import Foundation
import LeCodesSDK

public final class PushService: ServiceInstance {

    let channel: ServiceChannel
    private weak var engine: LeCodesEngine?

    public init(_ channel: ServiceChannel, _ engine: LeCodesEngine?) {
        self.channel = channel
        self.engine = engine
        PushManager.shared.attach(self)
    }

    public func call(_ method: String, _ args: [Any], _ settle: ChannelSettle) {
        let uuid = engine?.currentProjectUuid
        switch method {
        case "getStatus":
            // Works without an identity (permission is host-wide; registered = false) so the
            // SDK's listener-pipe poke never rejects.
            PushManager.shared.getStatus(uuid, settle)
        case "register":
            PushManager.shared.register(uuid, args.first as? [String: Any], settle)
        case "unregister":
            PushManager.shared.unregister(uuid, settle)
        case "getLaunch":
            PushManager.shared.getLaunch(uuid, settle)
        default:
            settle.reject("Unknown method: \(method)")
        }
    }

    public func close() {
        PushManager.shared.detach(self)
    }
}
